### DISCUSSION TITLE
Difference between environment variable and output parameters

### DISCUSSION THREAD

**Discussion by nvincent-vossloh** [community/community#55294] [Actions] [answered]
### Select Topic Area

Question

### Body

Hello,

I'm trying to understand the difference between environment variables and output parameters (of a step).

I am mainly using output parameters:
```
echo 'my_param=my_value' >> $GITHUB_OUTPUT
```
But the syntax to use an output parameter in a later step is quite long
```
steps.<step-id>.outputs.my_param
```
I have noticed that individual steps can generate environment variable on the flow:
```
echo 'my_env_var=my_value' >> $GITHUB_ENV
```
And it is quite easy to use in later steps (`$my_env_var`).

So when should we use one over the other, are there any risks related to one or the other ?

It looks like the output parameters needs to be used when one wish to set a job output parameters

Here are my sources of informations:
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable
https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter
https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

Side question, I use the following syntax to output parameters and env variables which I find nice to use for inspecting values:
```
echo 'my_param=my_value' | tee $GITHUB_OUTPUT
echo 'my_env_var=my_value' | tee $GITHUB_ENV
```
Is it safe or a bad practice ? (I'm not storing secrets in params or env var).

Thanks,
Nicolas.

**Comment by airtower-luna** [comment, marked as answer]
> So when should we use one over the other, are there any risks related to one or the other ?

In most cases: Use an environment variable if you want the value to be available to *all processes* in following steps *of the same job*. Environment variables set via `GITHUB_ENV` don't carry over to other jobs.

> It looks like the output parameters needs to be used when one wish to set a job output parameters

Correct. That's the main use for outputs. The other is if you need closer control over where it's available (like, only to specific steps).

Unless you need one of the special properties of outputs, using an environment variable is usually more convenient. For understanding it might be useful to know the environment variables are (slightly simplified) an operating system feature, the Actions runner just sets the requested ones for you. Any normal shell or programming language can use them. Outputs are an Actions feature, and only available where Actions syntax is evaluated.

To avoid odd bugs, use environment variables through the syntax of the shell (or other programming language) you're using (and *not* `${{ env.VAR }}`) wherever possible.

> ```
> echo 'my_param=my_value' | tee $GITHUB_OUTPUT
> echo 'my_env_var=my_value' | tee $GITHUB_ENV
> ```
> 
> Is it safe or a bad practice ? (I'm not storing secrets in params or env var).

The perfectly fine if you want the name/value pairs to show up in the log (so usually it is, but it won't be if they contain anything secret). For setting variables the only thing that matters is that the name/value pair ends up in the right file in the right format. :slightly_smiling_face:

**Reply by nvincent-vossloh** [reply]
Thanks for your detailed answer ! :slightly_smiling_face:
